# Exploring eBay Car Sales Data

For this project, we will be cleaning and analyzing data about used car listings on the German eBay website's 'classifieds' section.

In [187]:
import numpy as np
import pandas as pd

In [188]:
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [189]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [190]:
autos.info(); autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Analyzing the data, most null objects are located in the notRepairedDamage, vehicleType, fuelType, nodel, gearbox columns.
The dataset is also sorted from highest to lowest amounts and contains 20 columns of mostly strings in CamelCase.

In [192]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [193]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'picture_number', 'postal_code',
       'last_seen']
autos.columns = new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  picture_number  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [194]:
# Changed names of columns to snakecase to simplify labels into shorter lowercase titles

In [195]:
autos.describe(include = 'all')


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  picture_number   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Looking at the data, columns that could be dropped are: seller, unrepaired damage, offer type. The columns that need investigating/cleaning are: price, odometer.

In [197]:
autos['price'] = autos['price'].str.replace("$","")

In [198]:
autos['price'] = autos['price'].str.replace(",","").astype(int)

In [ ]:
x

In [199]:
autos["odometer"] = autos["odometer"].str.replace("km", "")


In [200]:
autos["odometer"] = autos["odometer"].str.replace(",", "")

In [201]:
autos["odometer"] = autos["odometer"].astype(int)

In [202]:
autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)

In [203]:
autos["odometer_km"].unique().shape

(13,)

In [204]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [205]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [206]:
autos["odometer_km"].sort_index()

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int64

The column "odometer_km" has 13 unique values, 65% of which is 150000 km. 

In [207]:
autos["price"].unique().shape

(2357,)

In [208]:
autos["price"].describe().round(2)


count       50000.00
mean         9840.04
std        481104.38
min             0.00
25%          1100.00
50%          2950.00
75%          7200.00
max      99999999.00
Name: price, dtype: float64

Looking at the summary of the price column's data values, we can identify anomalies with the minimum and maximum values. Car prices are not usually free, or as high as 99 million, especially for used car listings.

In [209]:
autos["price"].value_counts()


0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [210]:
autos["price"].sort_index(ascending= False)

49999     1250
49998    22900
49997    13200
49996     1980
49995    24900
         ...  
4         1350
3         4350
2         8990
1         8500
0         5000
Name: price, Length: 50000, dtype: int64

In [211]:
autos.loc[(autos["price"] > 150000)]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
10500  2016-03-17 12:56:38                                        Porsche_991   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
17140  2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
18509  2016-03-27 18:43:02                                    Porsche_991_GT3   
20351  2016-03-22 17:47:59                                    BMW_Z8_roadster   
22673  2016-03-05 18:40:13  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
32840  2016-03-22 08:55:27  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
33638  2016-03-24 12:54:19                                 Porsche_911_GT3_RS   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
36818  2016-03-27 18:37:37                                        Porsche_991   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
45387  2016-03-18 16:41:22  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offer_type     price   abtest vehicle_type  registration_year  \
514    privat    Angebot    999999     test        kombi               2009   
2897   privat    Angebot  11111111     test    limousine               1973   
7814   privat    Angebot   1300000  control        coupe               1992   
10500  privat    Angebot    155000     test        coupe               2013   
11137  privat    Angebot  10000000  control        coupe               1960   
11433  privat    Angebot    175000     test        coupe               2016   
12682  privat    Angebot    265000  control        coupe               2016   
14715  privat    Angebot    345000  control       cabrio               2012   
17140  privat    Angebot    180000     test       cabrio               2012   
18509  privat    Angebot    169000  control        coupe               2013   
20351  privat    Angebot    190000  control       cabrio               2000   
22673  privat    Angebot    163991     test        coupe               2014   
22947  privat    Angebot   1234566  control        kombi               1999   
24384  privat    Angebot  11111111     test          NaN               1

In [212]:
autos[autos["price"].between(100, 150000)] # Remove outliers

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  picture_number  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 1

After evaluating prices above 150000, several outliers were found within the price column values (including values from 10-99 million).
The value '0' occurs in 2.8% of the remaining data, which still proves to be an anomaly for car listing prices.

In [213]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [214]:
autos['date_crawled'].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 50000, dtype: object

In [215]:
autos["date_crawled"].value_counts(normalize = True, dropna = False)

2016-03-08 10:40:35    0.00006
2016-03-12 16:06:22    0.00006
2016-03-25 19:57:10    0.00006
2016-03-21 16:37:21    0.00006
2016-03-10 15:36:24    0.00006
                        ...   
2016-03-11 23:36:24    0.00002
2016-04-04 16:42:37    0.00002
2016-03-16 20:40:13    0.00002
2016-04-03 23:37:35    0.00002
2016-03-09 23:53:21    0.00002
Name: date_crawled, Length: 48213, dtype: float64

In [216]:
autos["date_crawled"].value_counts(normalize = True, dropna = False).sort_index().unique()

array([2.e-05, 4.e-05, 6.e-05])

Most percentages are either 0.00006, 0.00004, 0.00002 for dates, and the most common dates are in the month of March, which have occurrences of higher percentages like 0.00006.

In [217]:
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11 00:00:00    0.00002
2015-08-10 00:00:00    0.00002
2015-09-09 00:00:00    0.00002
2015-11-10 00:00:00    0.00002
2015-12-05 00:00:00    0.00002
                        ...   
2016-04-03 00:00:00    0.03892
2016-04-04 00:00:00    0.03688
2016-04-05 00:00:00    0.01184
2016-04-06 00:00:00    0.00326
2016-04-07 00:00:00    0.00128
Name: ad_created, Length: 76, dtype: float64

In [218]:
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index().unique()

array([2.000e-05, 4.000e-05, 6.000e-05, 8.000e-05, 1.200e-04, 2.000e-04,
       1.600e-04, 1.000e-04, 8.600e-04, 1.440e-03, 2.304e-02, 1.512e-02,
       3.474e-02, 3.334e-02, 3.324e-02, 3.186e-02, 3.278e-02, 3.662e-02,
       1.692e-02, 3.522e-02, 3.374e-02, 3.000e-02, 3.120e-02, 1.372e-02,
       3.384e-02, 3.786e-02, 3.772e-02, 3.280e-02, 3.218e-02, 2.908e-02,
       3.188e-02, 3.256e-02, 3.090e-02, 3.496e-02, 3.414e-02, 3.344e-02,
       3.192e-02, 3.380e-02, 3.508e-02, 3.892e-02, 3.688e-02, 1.184e-02,
       3.260e-03, 1.280e-03])

Larger percentages, going above 0.0002 start at the beginning of April. All dates have the time at 00:00:00. Highest percentage is 0.03892 on April 3rd.

In [219]:
autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
                        ...   
2016-04-07 14:58:44    0.00006
2016-04-07 14:58:45    0.00002
2016-04-07 14:58:46    0.00002
2016-04-07 14:58:48    0.00006
2016-04-07 14:58:50    0.00008
Name: last_seen, Length: 39481, dtype: float64

In [220]:
autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index().unique()

array([2.0e-05, 4.0e-05, 6.0e-05, 1.0e-04, 8.0e-05, 1.2e-04, 1.4e-04,
       1.6e-04])

In [221]:
autos["last_seen"].sort_index().unique()

array(['2016-04-06 06:45:54', '2016-04-06 14:45:08',
       '2016-04-06 20:15:37', ..., '2016-04-02 14:18:02',
       '2016-04-04 11:47:27', '2016-04-06 21:18:48'], dtype=object)

Percentages above 0.00002 appear most often in the month of April. The month of April has percentages ranging from 0.00002 to 0.00016.

In [222]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Looking at the summary for the registration_year data, two anomalies can be found in the minimum and mazimum values for the year values. It is unlikely in a data set with an average of the year 2005 occurring within it would include the year 1000 and 9999 as having car listings.

In [175]:
autos["registration_year"].value_counts(ascending=True).head(60)

1952     1
9996     1
1001     1
1929     1
1931     1
1948     1
1500     1
1000     1
8888     1
2800     1
1938     1
1939     1
5911     1
1943     1
4500     1
6200     1
1927     1
1111     1
1953     1
4800     1
4100     1
1934     2
1951     2
1941     2
1957     2
1800     2
1954     2
9000     2
1955     2
1950     3
2019     3
1937     4
1962     4
1958     4
9999     4
5000     4
1956     5
1961     6
1959     7
1963     9
1910     9
1964    12
1965    17
1969    19
1975    19
1977    22
1966    22
1974    24
1973    26
1968    26
1976    27
1967    27
1971    27
1981    31
1960    34
1979    35
1972    35
1982    43
1970    45
1978    47
Name: registration_year, dtype: int64

The highest and lowest values are: 2016 and 1900. The decades after 1900 had at least 1 or more registrations recorded within the data, and came after the 1900s. 2016 would be the maximum value since this data has car listings seen up to the year 2016.

In [223]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [228]:
autos["registration_year"].value_counts(normalize= True)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

The year with the most amount of registration is 2000, and registration declines after or before the year 2000, with some exceptions like 1999.

In [230]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [235]:
autos["brand"].describe()

count          48028
unique            40
top       volkswagen
freq           10188
Name: brand, dtype: object

In [231]:
autos["brand"].value_counts().head(20)

volkswagen        10188
bmw                5284
opel               5195
mercedes_benz      4580
audi               4149
ford               3352
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
citroen             669
smart               668
toyota              599
sonstige_autos      526
hyundai             473
volvo               444
mini                415
Name: brand, dtype: int64

10188 brands are Volkswagen, with BMW and Opel having similar values of 5284 and 5195, respectively. For my aggregate data, I will be using the top 20 brands as summarized above.

In [261]:
import pprint
brand_prices = {}
brands = autos["brand"].value_counts()[:20].index

for b in brands:
    selected = autos[autos["brand"] == b]
    mean = selected["price"].mean()
    brand_prices[b] = mean
formatted = []
for key in brand_prices:
    formatted.append((brand_prices[key], key))
formatted = sorted(formatted, reverse = True)
for value in formatted:
    print("{}: {}".format(value[1], value[0]))


citroen: 44534.79671150971
sonstige_autos: 39621.77946768061
mercedes_benz: 30317.447816593885
mini: 10460.012048192772
audi: 9093.65003615329
bmw: 8334.645155185466
ford: 7263.015811455847
volkswagen: 6516.457597173145
skoda: 6334.91948051948
hyundai: 5308.53911205074
opel: 5252.61655437921
toyota: 5115.33388981636
volvo: 4757.108108108108
nissan: 4664.891034482758
seat: 4296.492554410081
mazda: 4010.7716643741405
smart: 3542.706586826347
peugeot: 3039.4682651622
fiat: 2711.8011272141707
renault: 2395.4164467897976


The results show that the brands Citroen, Sonstige Autos, Mercedes Benz, and Mini gave the highest mean price, despite all except 1 not having the most occurrence in the data. Of the top 6 brands, Mercedes Benz is the only brand with a mean price above 10000. There is a large price range between the top 6: Audi, BMW, and Mercedes Benz are more expensive, Opel and Ford are less expensive, and Volkswagen is in between. A general correlation can be found between occurrence of the brand and its mean price: the higher the resale mean price, the fewer amount of cars occurring within the data. The inverse would then also be true: the lower the mean price, the higher occurrence within the data.

In [260]:
brand_mileage = {}
for b in brands:
    selected_2 = autos[autos["brand"] == b]
    miles = selected_2["odometer_km"].mean()
    brand_mileage[b] = miles
price_series = pd.Series(brand_prices)
df = pd.DataFrame(price_series, columns=["mean_price"])
df["mean_mileage"] = pd.Series(brand_mileage)
pprint.pprint(df)

                  mean_price   mean_mileage
volkswagen       6516.457597  128730.369062
bmw              8334.645155  132434.708554
opel             5252.616554  129227.141482
mercedes_benz   30317.447817  130860.262009
audi             9093.650036  129287.780188
ford             7263.015811  124046.837709
renault          2395.416447  128183.817062
peugeot          3039.468265  127136.812412
fiat             2711.801127  116553.945250
seat             4296.492554  121563.573883
skoda            6334.919481  110954.545455
mazda            4010.771664  124745.529574
nissan           4664.891034  118572.413793
citroen         44534.796712  119461.883408
smart            3542.706587   99595.808383
toyota           5115.333890  115709.515860
sonstige_autos  39621.779468   87262.357414
hyundai          5308.539112  106511.627907
volvo            4757.108108  138355.855856
mini            10460.012048   88602.409639


From the aggregate data, there seems to be no direct correlation between mileage or mean price. Citroen, with the highest mean price, has a mean mileage of around 119461, while Opel, with a considerably lower mean price, has a mean mileage of around 129227. This mileage is close to that of Mercedes Benz, which has a mean price of around 30317 and a mileage of 130860, compared to Opel's price of 5252. This indicates that even lower prices can have high mileage averages, and overall most of the data has mileages of 100000 and above that are close to another, showing that there is no correlation between price and mileage.